flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -2249,14 +2305,585 @@ class ScalarExprLowering {
+        expSpace{esp} {}
+
+  template <typename A>
+  ExtValue lower(const A &x) {
+    return gen(x);
+  }
+
+  template <typename A>
+  ExtValue lowerRef(const A &x) {
+    semant = ConstituentSemantics::RefOpaque;
+    return gen(x);
+  }
+
+  template <typename A, typename B>
+  ExtValue assign(const A &lhs, const B &rhs) {
+    semant = ConstituentSemantics::RefTransparent;
+    // 1) Lower the rhs expression with array_fetch op(s).
+    auto rexv = lower(rhs);
+    elementalValue = fir::getBase(rexv);
+    // 2) Lower the lhs expression to an array_update.
+    semant = ConstituentSemantics::ProjectedCopyInCopyOut;
+    auto lexv = lower(lhs);
+    // 3) Thread the array value updated forward
+    auto oldInnerArg =
+        mlir::cast<fir::ArrayUpdateOp>(fir::getBase(lexv).getDefiningOp())
+            .sequence();
+    auto offset = expSpace.argPosition(oldInnerArg);
+    expSpace.setInnerArg(offset, fir::getBase(lexv));
+    return lexv;
+  }
+
+private:
+  bool pathIsEmpty() { return reversePath.empty(); }
+
+  void clearPath() { reversePath.clear(); }
+
+  llvm::SmallVector<mlir::Value> lowerPath(mlir::Type ty) {
+    auto loc = getLoc();
+    auto fieldTy = fir::FieldType::get(builder.getContext());
+    auto idxTy = builder.getIndexType();
+    llvm::SmallVector<mlir::Value> result;
+    for (const auto &v : llvm::reverse(reversePath)) {
+      auto addField = [&](const Fortran::evaluate::Component &x) {
+        // TODO: Move to a helper function.
+        auto name = toStringRef(x.GetLastSymbol().name());
+        auto recTy = ty.cast<fir::RecordType>();
+        auto memTy = recTy.getType(name);
+        auto fld = builder.create<fir::FieldIndexOp>(
+            loc, fieldTy, name, recTy, /*typeparams=*/mlir::ValueRange{});
+        result.push_back(fld);
+        return memTy;
+      };
+      auto addSub = [&](const Fortran::evaluate::Subscript &sub) {
+        auto v = fir::getBase(gen(sub));
+        auto cast = builder.createConvert(loc, idxTy, v);
+        result.push_back(cast);
+      };
+      std::visit(Fortran::common::visitors{
+                     [&](int) { ty = fir::unwrapSequenceType(ty); },
+                     [&](const Fortran::evaluate::Subscript *x) { addSub(*x); },
+                     [&](const Fortran::evaluate::ArrayRef *x) {
+                       assert(!x->base().IsSymbol());
+                       ty = addField(x->base().GetComponent());
+                       for (const auto &sub : x->subscript())
+                         addSub(sub);
+                       ty = fir::unwrapSequenceType(ty);
+                     },
+                     [&](const Fortran::evaluate::Component *x) {
+                       ty = addField(*x);
+                     }},
+                 v);
+    }
+    return result;
+  }
+
+  /// Apply the reversed path components to the value returned from `load`.
+  ExtValue applyPathToArrayLoad(fir::ArrayLoadOp &load) {
```

Comment:
Should this be static?

Reply:
No. It's using member functions and member data.